c2rust-refactor/src/scripting/ast_visitor.rs

```diff
@@ -0,0 +1,655 @@
+use rlua::prelude::{LuaContext, LuaError, LuaFunction, LuaResult, LuaTable};
+use syntax::ast::{
+    Arg, BindingMode, Block, Crate, Expr, ExprKind, FunctionRetTy, FnDecl, ImplItem,
+    ImplItemKind, Item, ItemKind, Local, Mod, Mutability::*, NodeId, Pat, PatKind,
+    Stmt, StmtKind,
+};
+use syntax::source_map::symbol::Symbol;
+use syntax::ptr::P;
+
+use crate::ast_manip::fn_edit::{FnKind, FnLike};
+use crate::scripting::{IntoLuaAst, TransformCtxt, utils::iter_to_lua_array};
+
+macro_rules! call_lua_visitor_method {
+    ($obj: expr , $method: ident ($($params: expr),*)) => {
+        let opt_visit_method: Option<LuaFunction> = $obj.get(stringify!($method))?;
+
+        if let Some(visit_method) = opt_visit_method {
+            let proceed = visit_method.call::<_, bool>(($obj.clone(), $($params.clone()),*))?;
+
+            if !proceed {
+                return Ok(());
+            }
+        }
+    };
+}
+
+pub(crate) struct LuaAstVisitor<'a, 'lua, 'tctx: 'a> {
+    ctx: &'a TransformCtxt<'a, 'tctx>,
+    lua_ctx: LuaContext<'lua>,
+    visitor_obj: LuaTable<'lua>
+}
+
+impl<'a, 'lua, 'tctx> LuaAstVisitor<'a, 'lua, 'tctx> {
+    pub fn new(visitor_obj: LuaTable<'lua>, ctx: &'a TransformCtxt<'a, 'tctx>, lua_ctx: LuaContext<'lua>) -> Self {
+        LuaAstVisitor {
+            ctx,
+            lua_ctx,
+            visitor_obj,
+        }
+    }
+
+    pub fn visit_crate(&self, krate: &mut Crate) -> LuaResult<()> {
+        let lua_krate = krate.clone().into_lua_ast(self.ctx, self.lua_ctx)?;
+
+        call_lua_visitor_method!(self.visitor_obj,visit_crate(lua_krate));
+
+        self.visit_module(lua_krate.get("module")?)?;
+        self.finish()?;
+
+        krate.merge_lua_ast(lua_krate)?;
+
+        Ok(())
+    }
+
+    pub fn visit_module(&self, module: LuaTable<'lua>) -> LuaResult<()> {
```

Comment:
```suggestion
    pub fn visit_mod(&self, module: LuaTable<'lua>) -> LuaResult<()> {
```
I'd like to match the API in the libsyntax `Visitor` when it makes sense.